File: Labs/week7/A_build_tree_preorder.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

int n, idx = 0;
vector<string> seq;

TreeNode* build() {
    if (idx >= n || seq[idx] == "#") {
        idx++;
        return nullptr;
    }
    TreeNode* node = new TreeNode(stoi(seq[idx]));
    idx++;
    node->left = build();
    node->right = build();
    return node;
}

void preorder(TreeNode* root) {
    if (!root) return;
    cout << root->val << endl;
    preorder(root->left);
    preorder(root->right);
}

int main() {
    cin >> n;
    seq.resize(n);
    for (int i = 0; i < n; i++) cin >> seq[i];
    TreeNode* root = build();
    preorder(root);
    return 0;
}

File: Labs/week7/A_build_tree_preorder.py
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None

def build(seq, idx):
    if idx[0] >= len(seq) or seq[idx[0]] == "#":
        idx[0] += 1
        return None
    node = TreeNode(int(seq[idx[0]]))
    idx[0] += 1
    node.left = build(seq, idx)
    node.right = build(seq, idx)
    return node

def preorder(root):
    if not root:
        return
    print(root.val)
    preorder(root.left)
    preorder(root.right)

n = int(input())
seq = input().strip().split()
root = build(seq, [0])
preorder(root)
